src/Stratis.Bitcoin/Base/TipsManager.cs

```diff
@@ -0,0 +1,210 @@
+        /// commited such a tip would be able to recover on startup to it or any tip that is ancestor to tip commited.
+        /// </remarks>
+        void CommitTipPersisted(ITipProvider provider, ChainedHeader tip);
+    }
+
+    public class TipsManager : ITipsManager
+    {
+        private readonly IKeyValueRepository keyValueRepo;
+
+        private const string commonTipKey = "lastcommontip";
+
+        /// <summary>Highest commited tips mapped by their providers.</summary>
+        private readonly Dictionary<ITipProvider, ChainedHeader> tipsByProvider;
+
+        /// <summary>Highest tip commited between all registered components.</summary>
+        private ChainedHeader lastCommonTip;
+
+        /// <summary>Protects all access to <see cref="tipsByProvider"/> and write access to <see cref="lastCommonTip"/>.</summary>
+        private readonly object lockObject;
+
+        /// <summary>Triggered when <see cref="lastCommonTip"/> is updated.</summary>
+        private readonly AsyncManualResetEvent newCommonTipSetEvent;
+
+        private Task commonTipPersistingTask;
+
+        private readonly CancellationTokenSource cancellation;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        public TipsManager(IKeyValueRepository keyValueRepo, ILoggerFactory loggerFactory)
+        {
+            this.keyValueRepo = keyValueRepo;
+            this.tipsByProvider = new Dictionary<ITipProvider, ChainedHeader>();
+            this.lockObject = new object();
+            this.newCommonTipSetEvent = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <inheritdoc />
+        public void Initialize(ChainedHeader highestHeader)
+        {
+            if (this.commonTipPersistingTask != null)
+                throw new Exception("Already initialized.");
+
+            var commonTipHashHeight = this.keyValueRepo.LoadValue<HashHeightPair>(commonTipKey);
+
+            if (commonTipHashHeight != null)
+                this.lastCommonTip = highestHeader.FindAncestorOrSelf(commonTipHashHeight.Hash, commonTipHashHeight.Height);
+            else
+                // Genesis.
+                this.lastCommonTip = highestHeader.GetAncestor(0);
+
+            this.logger.LogDebug("Tips manager initialized at '{0}'.", this.lastCommonTip);
+
+            this.commonTipPersistingTask = this.PersistCommonTipContinuouslyAsync();
+        }
+
+        /// <summary>Continuously persists <see cref="lastCommonTip"/> to hard drive.</summary>
+        private async Task PersistCommonTipContinuouslyAsync()
+        {
+            while (!this.cancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    await this.newCommonTipSetEvent.WaitAsync(this.cancellation.Token).ConfigureAwait(false);
```

Comment:
feels very much like an async queue would do the job rather then an event sync

Reply:
Here we don't need a queue of items. 
So in case save takes 1 seconds and in 1 seconds we change common tip to 5a,6a,7a then we don't need to save them all one by one, only the most recent one.
That's why queue wasn't used but a reset event.